napari/plugins/_npe2.py

```diff
@@ -488,29 +504,183 @@ def _add_sample(
+            if widget_contribution := _npe2.get_widget_contribution(
+                mf.name,
+                widget_name,
+            ):
+                widget_callable, _ = widget_contribution
+                widget_param = _get_widget_param(widget_callable, widget_name)
+
+                kwargs = {}
+                if widget_param:
+                    kwargs[widget_param] = napari_viewer
+                return widget_callable(**kwargs), name
+            return None
+
+        def _get_current_dock_status(
+            window: Window,
+            name: str = full_name,
+        ):
+            if name in window._dock_widgets:
+                return window._dock_widgets[name].isVisible()
+            return False
+
+        title = full_name
+        if multiprovider:
+            title = widget.display_name
+        # To display '&' instead of creating a shortcut
+        title = title.replace("&", "&&")
+
+        widget_actions.append(
+            Action(
+                id=f'{mf.name}:{widget.display_name}',
+                title=title,
+                callback=_widget_callback,
+                menus=[
+                    {
+                        'id': submenu_id,
+                        'group': MenuGroup.PLUGIN_CONTRIBUTIONS,
+                    }
+                ],
+                toggled=ToggleRule(get_current=_get_current_dock_status),
+            )
+        )
+    return submenu, widget_actions
+
+
 def _register_manifest_actions(mf: PluginManifest) -> None:
     """Gather and register actions from a manifest.
 
     This is called when a plugin is registered or enabled and it adds the
     plugin's menus and submenus to the app model registry.
     """
     from napari._app_model import get_app
+    from napari._qt.qt_main_window import _QtMainWindow
 
     app = get_app()
     actions, submenus = _npe2_manifest_to_actions(mf)
     samples_submenu, sample_actions = _get_samples_submenu_actions(mf)
+    widgets_submenu, widget_actions = _get_widgets_submenu_actions(mf)
     context = pm.get_context(cast('PluginName', mf.name))
-    # Connect 'unregister' callback to plugin deactivate ('unregistered') event
+
+    # Register and connect dispose callback to plugin deactivate ('unregistered') event
+    actions = actions + sample_actions + widget_actions
     if actions:
         context.register_disposable(app.register_actions(actions))
+    submenus = submenus + samples_submenu + widgets_submenu
     if submenus:
         context.register_disposable(app.menus.append_menu_items(submenus))
-    if samples_submenu:
-        context.register_disposable(
-            app.menus.append_menu_items(samples_submenu)
-        )
-    if sample_actions:
-        context.register_disposable(app.register_actions(sample_actions))
+
+    # Register dispose functions that remove plugin widgets from widget dictionary
+    # `window._dock_widgets`
+    _qmainwin = _QtMainWindow.current()
```

Comment:
Would be nice to extract this also into it's own function if at all possible.

Reply:
It's actually defined here:

https://github.com/napari/napari/blob/e7f68dd4d0ee5b6fdbd1f3935de370db46e3ff5e/napari/_qt/_qapp_model/qactions/__init__.py#L42-L46

But I'm going to move these provider functions out so we can re-use them (and manually register in the init)

Reply:
Done, though I do wonder if these provider functions belong in `napari/_qt/_qapp_model/qactions/__init__.py`. The other providers are defined in `napari/_app_model/injection/_providers.py` but these seem to be non-qt providers.

Reply:
@lucyleeow yeah I thought the same thing when I came across them just then. But also yes agree that since they're `qt` dependent, they probably shouldn't go in `_providers.py`. Perhaps a `_qproviders.py` under `qactions`?

Reply:
Done, but def need to add a note to technical docs about the separation of qt/non-qt actions and providers.